### PR TITLE
Test EmbeddedSignatureData api

### PR DESCRIPTION
**Pull request by davidwrighton**
- Add test to provide clarity about the behavior of the EmbeddedSignatureData api when examining the start of the signature for custom modifiers

### CODE REVIEW

**Comment by davidwrighton** [comment]
@elinor-fung is about to make a change which will require testing for only the custommodifiers at the start of a signature. This test codifies the behavior of EmbeddedSignatureData in that circumstance.

**Comment by davidwrighton** [comment]
@dotnet/crossgen-contrib

**Review by mangod9 — APPROVED** [review]

File: src/coreclr/src/tools/aot/ILCompiler.TypeSystem.ReadyToRun.Tests/SignatureTests.cs [hunk 1]
```diff
@@ -27,6 +27,42 @@ public SignatureTests()
...
+        {
+            if (!signature.HasEmbeddedSignatureData || signature.GetEmbeddedSignatureData() == null)
+                return "";
+
+            StringBuilder sb = new StringBuilder();
+            foreach (EmbeddedSignatureData data in signature.GetEmbeddedSignatureData())
+            {
+                sb.Append(data.kind.ToString());
+                sb.Append(data.index);
+                sb.Append(data.type.ToString());
```
Comment (MichalStrehovsky) [inline src/coreclr/src/tools/aot/ILCompiler.TypeSystem.ReadyToRun.Tests/SignatureTests.cs:40]: `TypeDesc.ToString` is not considered stable. I would rather not introduce dependencies on it.

**Review by MichalStrehovsky — COMMENTED** [review]
Sigh. We have plenty of good APIs that don't have unit tests, but the one API that should be nuked from the orbit... that one we add a unit test for.

Still wondering how we're going to use this bad API to e.g. fix the bug where crossgen2 currently ignores copy constructors in marshalling:

https://github.com/dotnet/runtime/blob/6072e4d3a7a2a1493f514cdf4be75a3d56580e84/src/coreclr/src/vm/mlinfo.cpp#L1314-L1327

I think we should just bite the bullet delete this horrible API and replace it with a structural model that people can sanely query.

**Comment by davidwrighton** [comment]
While the api is in fact... strictly the worst I've ever had the displeasure of operating on, that's actually the best reason around to write a unit test, as its not really plausible that anyone will ever actually understand the darn thing intuitively, and for each case where we use the api really needs a unit test to ensure that particular use remains functional. For that particular marshalling concern, I think that the api is actually useable for that, as I think it will produce indices in a particular easily testable pattern. 

It will probably be something like...

`$"0.1.1.{parameterIndex + 2}.1"`

Since the test you're looking at has some extra BYREF/PTR scenarios, its a bit more complicated than that, but it should actually be workable.

**Comment by MichalStrehovsky** [comment]
> Since the test you're looking at has some extra BYREF/PTR scenarios, its a bit more complicated than that, but it should actually be workable.

Workable, yes. But I still question whether the time spent inventing the right "IP addresses" and proving they're correct in code reviews would be better spent designing a structural API to query these instead.

I filed #39423 for the p/invoke issue.

**Pull request merged**